### PR TITLE
Metrics: Handle null observations in observers

### PR DESCRIPTION
**Pull request by stu-elastic**
Avoid null pointer error when an observation is null [0].

* To get the tests working, replace the `this-escape` buildInstrument with a Builder.
* Removes locking around close

[0] 
```
Cannot invoke "org.elasticsearch.telemetry.metric.DoubleWithAttributes.value()" because "observation" is null
java.lang.NullPointerException: Cannot invoke "org.elasticsearch.telemetry.metric.DoubleWithAttributes.value()" because "observation" is null
   at org.elasticsearch.telemetry.apm@8.12.0-SNAPSHOT/org.elasticsearch.telemetry.apm.internal.metrics.DoubleAsyncCounterAdapter.lambda$buildInstrument$0(DoubleAsyncCounterAdapter.java:44)
   at co.elastic.apm.agent.opentelemetry.metrics.bridge.v1_14.BridgeDoubleCounterBuilder$1.accept(BridgeDoubleCounterBuilder.java:60)
   at co.elastic.apm.agent.opentelemetry.metrics.bridge.v1_14.BridgeDoubleCounterBuilder$1.accept(BridgeDoubleCounterBuilder.java:57)
   at co.elastic.apm.agent.embeddedotel.proxy.ProxyDoubleCounterBuilder$1.accept(ProxyDoubleCounterBuilder.java:60)
   at co.elastic.apm.agent.embeddedotel.proxy.ProxyDoubleCounterBuilder$1.accept(ProxyDoubleCounterBuilder.java:57)
   at io.opentelemetry.sdk.metrics.AbstractInstrumentBuilder.lambda$registerDoubleAsynchronousInstrument$0(AbstractInstrumentBuilder.java:109)
   at io.opentelemetry.sdk.metrics.internal.state.CallbackRegistration.invokeCallback(CallbackRegistration.java:84)
   at io.opentelemetry.sdk.metrics.internal.state.MeterSharedState.collectAll(MeterSharedState.java:96)
   at io.opentelemetry.sdk.metrics.SdkMeter.collectAll(SdkMeter.java:75)
   at io.opentelemetry.sdk.metrics.SdkMeterProvider$LeasedMetricProducer.produce(SdkMeterProvider.java:193)
   at io.opentelemetry.sdk.metrics.SdkMeterProvider$SdkCollectionRegistration.collectAllMetrics(SdkMeterProvider.java:217)
   at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.doRun(PeriodicMetricReader.java:161)
   at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.run(PeriodicMetricReader.java:153)
   at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
   at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
   at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
   at co.elastic.apm.agent.util.ExecutorUtils$2.run(ExecutorUtils.java:99)
   at java.base/java.lang.Thread.run(Thread.java:1583)
```

### CODE REVIEW

**Review by pgomulka — APPROVED** [review]
this looks really good.
approved, thanks

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/OtelHelper.java [hunk 1]
```diff
@@ -41,4 +48,36 @@ static Attributes fromMap(Map<String, Object> attributes) {
...
         return builder.build();
     }
+
+    static Consumer<ObservableDoubleMeasurement> doubleMeasurementCallback(Supplier<DoubleWithAttributes> observer) {
+        return measurement -> {
+            DoubleWithAttributes observation;
+            try {
+                observation = observer.get();
+            } catch (RuntimeException err) {
+                assert false : "observer must not throw [" + err.getMessage() + "]";
```
Comment (mosche) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/OtelHelper.java:58]: I see this code was just moved. Though, I'm a bit concerned about silently swallowing runtime exceptions if assertions are not explicitly enabled. That could  make it difficult to troubleshoot issues when metrics are missing.

**Review by mosche — COMMENTED** [review]
Nice small refactoring, looking good to me as well 👍 
My comments are mostly unrelated, nothing that needs to be tackled in this PR except the obsolete code.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/AbstractInstrument.java [hunk 2]
```diff
@@ -25,49 +26,51 @@
...
-        this.description = Objects.requireNonNull(description);
-        this.unit = Objects.requireNonNull(unit);
-        this.delegate = new AtomicReference<>(doBuildInstrument(meter));
-    }
-
-    private T doBuildInstrument(Meter meter) {
-        return AccessController.doPrivileged((PrivilegedAction<T>) () -> buildInstrument(meter));
+    public AbstractInstrument(Meter meter, Builder<T> builder) {
+        this.name = builder.getName();
+        this.instrumentBuilder = m -> AccessController.doPrivileged((PrivilegedAction<T>) () -> builder.build(m));
```
Comment (mosche) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/AbstractInstrument.java:35]: Sorry, I'm not really familiar with the security manager, forgive me my ignorance :/ I'll have to do some homework there! But wondering what change in this PR made it necessary to elevate permissions when constructing the instrument? 

Considering `AbstractInstrument` is part of the exports of the apm module and could be extended, e.g. in a plugin calling arbitrary user code in `builder.build(m)`, is it safe to run that with elevated permissions? What would be the context in such a case?

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/LongUpDownCounterAdapter.java [hunk 3]
```diff
@@ -24,13 +25,11 @@ public class LongUpDownCounterAdapter extends AbstractInstrument<LongUpDownCount
...
     public LongUpDownCounterAdapter(Meter meter, String name, String description, String unit) {
-        super(meter, name, description, unit);
+        super(meter, new Builder(name, description, unit));
     }
 
-    @Override
-    protected io.opentelemetry.api.metrics.LongUpDownCounter buildInstrument(Meter meter) {
-        var builder = Objects.requireNonNull(meter).upDownCounterBuilder(getName());
-        return builder.setDescription(getDescription()).setUnit(getUnit()).build();
+    protected static Function<Meter, LongUpDownCounter> instrumentBuilder(String name, String description, String unit) {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/LongUpDownCounterAdapter.java:31]: Yes.  Left over from the earlier refactor where the builders were lambdas.  I think the Builder class is more clear.  Will remove.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/DoubleAsyncCounterAdapter.java [hunk 4]
```diff
@@ -11,47 +11,40 @@
...
-            DoubleWithAttributes observation;
-            try {
-                observation = observer.get();
-            } catch (RuntimeException err) {
-                assert false : "observer must not throw [" + err.getMessage() + "]";
-                return;
-            }
-            measurement.record(observation.value(), OtelHelper.fromMap(observation.attributes()));
-        });
+    public void close() throws Exception {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/DoubleAsyncCounterAdapter.java:28]: That's the signature for the [AutoClosable](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html#close--) interface extended by `DoubleAsyncCounter`.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/LongUpDownCounterAdapter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/LongUpDownCounterAdapter.java]: Removed.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/OtelHelper.java (hunk 1 above)
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/OtelHelper.java:58]: We could add logging here.

Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/OtelHelper.java:62]: That should help, I think that's worth doing 👍

**Review by mosche — APPROVED** [review]
Thanks, LGTM 👍 

@stu-elastic / @pgomulka, happy to chat about the other unrelated comments in case you think these require a follow up.

**Pull request merged**